README.pod
=pod

=head1 NAME

Dancer2::Plugin::UnicodeNormalize - Normalize incoming Unicode parameters

=head1 VERSION

Version 0.03

=head1 SYNOPSIS

    use Dancer2::Plugin::UnicodeNormalize;

=head1 DESCRIPTION

Dancer2::Plugin::UnicodeNormalize normalizes all incoming parameters to a given
normalization form. This is achieved with a before hook, which should run
silently before processing each route. By default, we use Unicode Normalization
Form C - this is usually what you want. Other forms can be selected, see:
L</"CONFIGURATION">.

This plugin was inspired by L<Mojolicious::Plugin::UnicodeNormalize>. For
information on why Unicode Normalization is important, please see:

L<http://www.perl.com/pub/2012/05/perlunicookbook-unicode-normalization.html>

L<http://www.modernperlbooks.com/mt/2013/11/mojolicious-unicode-normalization-plugin-released.html>

=head1 CONFIGURATION

    plugins:
        UnicodeNormalize:
            form: NFC
            exclude:
                - '^/(css|javascripts|images)'

The C<form> parameter is described in L<Unicode::Normalize>. Default is NFC.

The C<exclude> parameter consists of a list of regular expressions to match
routes we do not wish to process parameters for.

=head1 AUTHOR

John Barrett, <john@jbrt.org>

=head1 CONTRIBUTING

L<http://github.com/jbarrett/Dancer2-Plugin-UnicodeNormalize>

All comments and contributions welcome.

=head1 BUGS AND SUPPORT

Please direct all requests to L<http://github.com/jbarrett/Dancer2-Plugin-UnicodeNormalize/issues>
or email <john@jbrt.org>.

=head1 COPYRIGHT

Copyright 2013 John Barrett.

=head1 LICENSE

This application is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

L<Dancer2>

L<Unicode::Normalize>

L<Mojolicious::Plugin::UnicodeNormalize>


=cut



Changes
Revision history for perl module Dancer2::Plugin::UnicodeNormalize

0.034 2016-06-09

  - Fixing build for Dancer 0.200000 (Peter Mottram, Sawyer X)

0.033 2014-12-11

  - Add minumum required Dancer 2 version and additional metadata
    (Alexandr Ciornii)

0.032 2014-10-30

  - Migrating tests to Plack::Test from deprecated Dancer2::Test

0.031 2013-11-28

  - Fix for perl 5.8

0.03 2013-11-27

  - Fixes for perls 5.10 - 5.14

0.02 2013-11-26

  - Fix for undefined parameter values - these are now skipped

0.01 2013-11-26

  - Initial release



t/lib/TestApp.pm
package t::lib::TestApp;

use Dancer2;
use Dancer2::Plugin::UnicodeNormalize;

set charset => 'UTF-8';
set apphandler => 'PSGI';
set log => 'error';

get '/cmp/:string1/:string2' => sub {
    return 'missing parameter!' unless ( param('string1') && param('string2') );
    return ( ( param('string1') eq param('string2') ) ? 'eq' : 'ne' );
};

post '/cmp' => sub {
    return 'missing parameter!' unless ( param('string1') && param('string2') );
    return ( ( param('string1') eq param('string2') ) ? 'eq' : 'ne' );
};

post '/upload' => sub {
    my ($file1, $file2) = (upload('file1'), upload('file2'));
    return ( ( $file1->content eq $file2->content ) ? 'eq' : 'ne' );
};

get '/form/:string' => sub {
    return param('string');
};

get '/optional/:string?' => sub {
    return 'success';
};

1;



lib/Dancer2/Plugin/UnicodeNormalize.pm
use strict;
use warnings;

package Dancer2::Plugin::UnicodeNormalize;
{
    $Dancer2::Plugin::UnicodeNormalize::VERSION = '0.034';
}

use Dancer2::Plugin;
use Unicode::Normalize;

on_plugin_import {
    my $dsl = shift;

    # pre Plugin2 we need to get settings here
    my $settings = plugin_setting;
    $dsl->app->add_hook(
        Dancer2::Core::Hook->new(
            name => 'before',
            code => sub {
                # perhaps better ways to check for Plugin2 but this works
                if ( $dsl->can('execute_plugin_hook')) {
                    # Plugin2  - fetch fresh config on each run
                    $settings = plugin_setting;
                }
                for (@{$settings->{'exclude'}}) { return if $dsl->app->request->path =~ /$_/ }

                my $form = $settings->{'form'} || 'NFC';
                my $normalizer = Unicode::Normalize->can($form);

                unless ($normalizer) {
                    require Carp;
                    Carp::croak( "Invalid normalization form '$form' requested" );
                }

                for (qw/query body route/) {
                    my $p = $dsl->app->request->params($_);
                    next unless $p;
                    %{$p} = map { $_ => $normalizer->($p->{$_}) } grep { $p->{$_} } keys %{$p};
                }
                $dsl->app->request->_build_params;
            },
        ),
    );
};

register_plugin for_versions => [2];

1;

__END__
=pod

=head1 NAME

Dancer2::Plugin::UnicodeNormalize - Normalize incoming Unicode parameters

=head1 VERSION

Version 0.034

=head1 SYNOPSIS

    use Dancer2::Plugin::UnicodeNormalize;

=head1 DESCRIPTION

Dancer2::Plugin::UnicodeNormalize normalizes all incoming parameters to a given
normalization form. This is achieved with a before hook, which should run
silently before processing each route. By default, we use Unicode Normalization
Form C - this is usually what you want. Other forms can be selected, see:
L</"CONFIGURATION">.

B<** NOTE **> This has been updated to build against Dancer2 0.200000 but
operations do not yet apply to newer parameter access methods, query_parameters,
body_parameters and route_parameters. This will be fixed in the next release.


This plugin was inspired by L<Mojolicious::Plugin::UnicodeNormalize>. For
information on why Unicode Normalization is important, please see:

L<http://www.perl.com/pub/2012/05/perlunicookbook-unicode-normalization.html>

L<http://www.modernperlbooks.com/mt/2013/11/mojolicious-unicode-normalization-plugin-released.html>

=head1 CONFIGURATION

    plugins:
        UnicodeNormalize:
            form: NFC
            exclude:
                - '^/(css|javascripts|images)'

The C<form> parameter is described in L<Unicode::Normalize>. Default is NFC.

The C<exclude> parameter consists of a list of regular expressions to match
routes we do not wish to process parameters for.

=head1 AUTHOR

John Barrett, <john@jbrt.org>

=head1 CONTRIBUTING

L<http://github.com/jbarrett/Dancer2-Plugin-UnicodeNormalize>

All comments and contributions welcome.

=head1 BUGS AND SUPPORT

Please direct all requests to L<http://github.com/jbarrett/Dancer2-Plugin-UnicodeNormalize/issues>
or email <john@jbrt.org>.

=head1 COPYRIGHT

Copyright 2013 John Barrett.

=head1 LICENSE

This application is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

L<Dancer2>

L<Unicode::Normalize>

L<Mojolicious::Plugin::UnicodeNormalize>

=cut

